qe/drivers.py

```diff
@@ -2019,29 +1974,54 @@ def E_pt2(self, psi_coef: Psi_coef) -> Energy:
 def selection_step(
     comm, lewis: Hamiltonian_generator, n_ord, psi_coef: Psi_coef, psi_det: Psi_det, n
 ) -> Tuple[Energy, Psi_coef, Psi_det]:
-    # 1. Generate a list of all the external determinant and their pt2 contribution
-    # 2. Take the n  determinants who have the biggest contribution and add it the wave function psi
+    # 1. Each MPI rank has a chunk of external determinants and computes their E_pt2 contribution
+    # 2. Take the n determinants who have the biggest contribution and add it the wave function psi
     # 3. Diagonalize H corresponding to this new wave function to get the new variational energy, and new psi_coef.
 
     # In the main code:
     # -> Go to 1., stop when E_pt2 < Threshold || N < Threshold
     # See example of chained call to this function in `test_f2_631g_1p5p5det`
     # 1.
-    psi_external_energy = Powerplant_manager(comm, lewis).psi_external_pt2(psi_coef)
+    PP_manager = Powerplant_manager(comm, lewis)
+    # Compute E_pt2 contributions of the distributed external space
+    psi_external_energy = PP_manager.psi_external_pt2(psi_coef)
 
     # 2.
-    idx = np.argpartition(psi_external_energy, n)[:n]
-    psi_external_det = Powerplant_manager(comm, lewis).psi_external
-    psi_det_extented = psi_det + [psi_external_det[i] for i in idx]
+    # Get indices of top n determinants from each rank + associated dets
+    # E_pt2 < 0, so n smallest are actually the largest contributors
+    local_idx = np.argpartition(psi_external_energy, n)[:n]
+    # External dets are ordered consistent w/ local_idx
+    local_dets = [PP_manager.psi_external_chunk[i] for i in local_idx]
+    # MPI Gather call + partial sort to get n best of these indices
+    # Ugly, but at least we can store this in memory
+    local_energies = np.array(psi_external_energy[local_idx], dtype="float")
+    aggregated_energies = np.zeros((comm.Get_size()) * n, dtype="float")
+    # Gather best local contributions + associated determinants
+    comm.Allgather([local_energies, MPI.DOUBLE], [aggregated_energies, MPI.DOUBLE])
+    aggregated_dets = comm.allgather(local_dets)  # Lower-case command for python objs
+    # Partial sort the gathered contributions to get the global n biggest E_pt2 contributions
+    if comm.Get_size() > 1:  # Else, out of index error is thrown
+        global_idx = np.argpartition(aggregated_energies, n)[:n]
+    else:
+        global_idx = np.arange(n)
+
+    # Add new determinants to the internal wavefunction
+    psi_det_extented = psi_det
+    for i in global_idx:
+        # rank_loc gives the rank where determinant with index i in idx is located
+        # index is the index of that external determinant on rank_loc
+        rank_loc, det_index = divmod(i, n)
+        local_list = aggregated_dets[rank_loc]  # external determinants on MPI rank: rank_loc
+        psi_det_extented = psi_det_extented + [local_list[det_index]]
```

Comment:
Maybe this can be simplify with something like this:
```
>>> E = np.array([0.2,0.3,0.1,0.8])
>>> D = np.array(["A","B","C","D"])
>>> idx = np.argpartition(-E,2)
>>> E[idx][:2]
array([0.8, 0.3])
>>> D[idx][:2]
array(['D', 'B'], dtype='<U1')
```